File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        visited = {}
        l = 0
        ans = 0
        for r in range(len(s)):
            if s[r] not in visited:
                ans = max(ans,r-l+1)
            else:
                if visited[s[r]]<l:
                    ans = max(ans,r-l+1)
                else:
                    l = visited[s[r]]+1
            visited[s[r]] = r
        return ans
    

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> temp;
        int l = 0;
        int r = 0;
        int count = 0;
        int ans = 0;
        while(r<s.size()){
            if(temp.find(s[r]) == temp.end() || temp[s[r]] == 0){
                count++;
                temp[s[r]] = 1;
                ans = max(ans,count);
                r++;
            }else{
                temp[s[l]]--;
                l++;
                count--;
            }
        }
        return ans;
    }
};